packages/engine/src/datastore/batch/mod.rs

```diff
@@ -9,105 +9,405 @@ pub mod message;
-        debug_assert!(self.memory().validate_markers());
-        let rb_msg = load::record_batch_message(self)?;
-        *self.record_batch_mut() = load::record_batch(self, rb_msg, self.record_batch().schema())?;
-        Ok(())
+/// Batch with Arrow data that can be accessed as an Arrow record batch
+pub struct ArrowBatch {
+    segment: Segment,
+
+    /// Arrow `RecordBatch` with references to `self.segment`
+    rb: RecordBatch,
+
+    /// Metadata referring to positions, sizes, null counts
+    /// and value counts of different Arrow buffers
+    dynamic_meta: DynamicMeta,
+
+    /// Map of which Arrow `Buffer`s and `FieldNode`s
+    /// correspond to which column
+    static_meta: Arc<StaticMeta>,
+
+    /// When growable columns are modified, their Arrow intermediate
+    /// column representations are kept here and wait for the
+    /// `self.flush_changes()` call, which inserts them into
+    /// `self.segment`.
+    changes: Vec<ColumnChange>,
+
+    /// Currently loaded version of memory and batch
+    loaded: Metaversion,
+}
+
+impl Deref for ArrowBatch {
+    type Target = Segment;
+
+    fn deref(&self) -> &Self::Target {
+        &self.segment
     }
+}
 
-    fn record_batch(&self) -> &RecordBatch;
-    fn record_batch_mut(&mut self) -> &mut RecordBatch;
+impl DerefMut for ArrowBatch {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.segment
+    }
 }
 
-pub trait DynamicBatch: ArrowBatch {
-    fn push_change(&mut self, change: ArrayChange) -> Result<()>;
-    fn flush_changes(&mut self) -> Result<()>;
+impl GrowableBatch<ArrayData, ColumnChange> for ArrowBatch {
+    fn static_meta(&self) -> &StaticMeta {
+        &self.static_meta
+    }
+
+    fn dynamic_meta(&self) -> &DynamicMeta {
+        &self.dynamic_meta
+    }
+
+    fn dynamic_meta_mut(&mut self) -> &mut DynamicMeta {
+        &mut self.dynamic_meta
+    }
+
+    fn memory_mut(&mut self) -> &mut Memory {
+        self.segment.memory_mut()
+    }
+
+    fn memory(&self) -> &Memory {
+        self.segment.memory()
+    }
+}
+
+impl ArrowBatch {
+    /// Add change to internal queue without checking metaversions.
+    fn queue_change_unchecked(&mut self, change: ColumnChange) -> Result<()> {
+        self.changes.push(change);
+        Ok(())
+    }
+
+    /// Return record batch without checking whether the
+    /// latest persisted data has been loaded.
+    fn record_batch_unchecked(&self) -> &RecordBatch {
+        &self.rb
+    }
+
+    /// Return record batch without checking whether the
+    /// latest persisted data has been loaded.
+    fn record_batch_unchecked_mut(&mut self) -> &mut RecordBatch {
+        &mut self.rb
+    }
+
+    fn loaded_metaversion_mut(&mut self) -> &mut Metaversion {
+        &mut self.loaded
+    }
 
+    /// Reload record batch (without checking metaversions).
+    fn reload_record_batch(&mut self) -> Result<()> {
+        debug_assert!(self.memory().validate_markers());
```

Comment:
Could you add a description to the assertion?

Reply:
Maybe I should add `_unchecked` here too